## UDEMY COURSES DATASET
This dataset contains all course data for all subjects.

We will analyze this data using the Pandas Library.

Here, the quesitons are given for the practice and learning purpose.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Udemy_dataset7.csv')
df[:5]

course_id                                       course_title  is_paid  \
0     288942  #1 Piano Hand Coordination: Play 10th Ballad i...     True   
1    1170074  #10 Hand Coordination - Transfer Chord Ballad ...     True   
2    1193886  #12 Hand Coordination: Let your Hands dance wi...     True   
3    1116700  #4 Piano Hand Coordination: Fun Piano Runs in ...     True   
4    1120410  #5  Piano Hand Coordination:  Piano Runs in 2 ...     True   

  price  num_subscribers  num_reviews  num_lectures               level  \
0    35             3137           18            68          All Levels   
1    75             1593            1            41  Intermediate Level   
2    75              482            1            47  Intermediate Level   
3    75              850            3            43  Intermediate Level   
4    75              940            3            32  Intermediate Level   

  content_duration   published_timestamp              subject  
0        1.5 hours  2014-09-18T05:07:05Z  Musical Instruments  
1           1 hour  2017-04-12T19:06:34Z  Musical Instruments  
2        1.5 hours  2017-04-26T18:34:57Z  Musical Instruments  
3           1 hour  2017-02-21T23:48:18Z  Musical Instruments  
4          37 mins  2017-02-21T23:44:49Z  Musical Instruments

## 1. What are all different subjects for which Udemy is offering courses ?

In [3]:
df.subject.unique()

array(['Musical Instruments', 'Business Finance', 'Graphic Design',
       'Web Development'], dtype=object)

## 2. Which subject has the maximum number of courses.

In [4]:
df.subject.value_counts()

Web Development        1200
Business Finance       1199
Musical Instruments     680
Graphic Design          603
Name: subject, dtype: int64

Web development -1200 and Business Finance- 1199 are having maximum number of courses

## 3. Show all the courses which are Free of Cost

### As we will find the number pf courses so we need to see whether there is any duplicate rows

In [5]:
df.duplicated().sum()

6

In [6]:
df.drop_duplicates(inplace=True)

In [7]:
df[:2]

course_id                                       course_title  is_paid  \
0     288942  #1 Piano Hand Coordination: Play 10th Ballad i...     True   
1    1170074  #10 Hand Coordination - Transfer Chord Ballad ...     True   

  price  num_subscribers  num_reviews  num_lectures               level  \
0    35             3137           18            68          All Levels   
1    75             1593            1            41  Intermediate Level   

  content_duration   published_timestamp              subject  
0        1.5 hours  2014-09-18T05:07:05Z  Musical Instruments  
1           1 hour  2017-04-12T19:06:34Z  Musical Instruments

In [8]:
# we have is_paid column which describes course is free or paid

df[df.is_paid == False]

course_id                                       course_title  is_paid  \
41       286070        5 lecciones que todo guitarrista debe tomar    False   
49       696630  7 Ways A Beginner Guitarist Can Sound Better, ...    False   
57       955914  A beginner`s guide to fingerpicking and strumm...    False   
67       270976                             A how to guide in HTML    False   
74      1214144  ¡Triunfar en La Bolsa de Valores No Requiere d...    False   
...         ...                                                ...      ...   
3620     552598  WordPress in under an hour - Create/manage you...    False   
3638    1068590                     Wordpress Website Mastery 2017    False   
3652     752706  Write quicker HTML5 and CSS 3; productivity ha...    False   
3674    1257976  Your First 10 Guitar Lessons - Learn how to pl...    False   
3680    1225194  Zoho Books Gestion Financière d'Entreprise pas...    False   

     price  num_subscribers  num_reviews  num_lectures           level  \
41    Free             4452          263            14  Beginner Level   
49    Free             4529          193             7  Beginner Level   
57    Free             3481           29            20  Beginner Level   
67    Free             7318          205             8  Beginner Level   
74    Free              338            7             6  Beginner Level   
...    ...              ...          ...           ...             ...   
3620  Free             6474           86            15      All Levels   
3638  Free             5181           64            22  Beginner Level   
3652  Free            13016          286            15      All Levels   
3674  Free              924            4            18  Beginner Level   
3680  Free              229            0            33      All Levels   

     content_duration   published_timestamp              subject  
41             1 hour  2014-08-23T05:08:14Z  Musical Instruments  
49            36 mins  2015-12-21T18:50:50Z  Musical Instruments  
57            2 hours  2016-09-13T21:51:59Z  Musical Instruments  
67            35 mins  2014-08-10T20:19:10Z      Web Development  
74             1 hour  2017-05-30T14:30:12Z     Business Finance  
...               ...                   ...                  ...  
3620           1 hour  2015-07-15T00:21:56Z      Web Development  
3638        2.5 hours  2017-01-18T17:01:40Z      Web Development  
3652        1.5 hours  2016-03-03T17:45:31Z      Web Development  
3674           1 hour  2017-06-29T00:29:36Z  Musical Instruments  
3680          2 hours  2017-05-26T16:45:55Z     Business Finance  

[310 rows x 11 columns]

#### There are 310 Free courses

## 4. Show all the courses which are paid

In [9]:
df[df.is_paid == True]

course_id                                       course_title  is_paid  \
0        288942  #1 Piano Hand Coordination: Play 10th Ballad i...     True   
1       1170074  #10 Hand Coordination - Transfer Chord Ballad ...     True   
2       1193886  #12 Hand Coordination: Let your Hands dance wi...     True   
3       1116700  #4 Piano Hand Coordination: Fun Piano Runs in ...     True   
4       1120410  #5  Piano Hand Coordination:  Piano Runs in 2 ...     True   
...         ...                                                ...      ...   
3676     498488  Your First Successful Forex Trades - With Case...     True   
3677     328960  Your Own Site in 45 Min: The Complete Wordpres...     True   
3678     552700    Your Second Course on Piano: Two Handed Playing     True   
3679     631754  Zend Framework 2: Learn the PHP framework ZF2 ...     True   
3681     964478                Zombie Apocalypse Photoshop Actions     True   

     price  num_subscribers  num_reviews  num_lectures               level  \
0       35             3137           18            68          All Levels   
1       75             1593            1            41  Intermediate Level   
2       75              482            1            47  Intermediate Level   
3       75              850            3            43  Intermediate Level   
4       75              940            3            32  Intermediate Level   
...    ...              ...          ...           ...                 ...   
3676   200             1079           34            16          All Levels   
3677   120             1566           29            36          All Levels   
3678    70             1018           12            22      Beginner Level   
3679    40              723          130            37          All Levels   
3681    50               12            1            15          All Levels   

     content_duration   published_timestamp              subject  
0           1.5 hours  2014-09-18T05:07:05Z  Musical Instruments  
1              1 hour  2017-04-12T19:06:34Z  Musical Instruments  
2           1.5 hours  2017-04-26T18:34:57Z  Musical Instruments  
3              1 hour  2017-02-21T23:48:18Z  Musical Instruments  
4             37 mins  2017-02-21T23:44:49Z  Musical Instruments  
...               ...                   ...                  ...  
3676        2.5 hours  2015-05-26T20:48:48Z     Business Finance  
3677          4 hours  2015-04-20T22:15:17Z      Web Development  
3678          5 hours  2015-10-26T20:04:21Z  Musical Instruments  
3679        6.5 hours  2015-11-11T18:55:45Z      Web Development  
3681        1.5 hours  2016-09-26T22:19:48Z       Graphic Design  

[3366 rows x 11 columns]

#### There are 3366 paid courses

## 5. Which are top selling Courses?

In [10]:
#To Find Top Selling course we must see the number of subscribers

df.num_subscribers

0       3137
1       1593
2        482
3        850
4        940
        ... 
3677    1566
3678    1018
3679     723
3680     229
3681      12
Name: num_subscribers, Length: 3676, dtype: int64

In [11]:
df.sort_values('num_subscribers',ascending=False).head(10)

course_id                                       course_title  is_paid  \
2230      41295               Learn HTML5 Programming From Scratch    False   
776       59014                     Coding for Entrepreneurs Basic    False   
3385     625204                         The Web Developer Bootcamp     True   
640      173548  Build Your First Website in 1 Week with HTML5 ...    False   
3316     764164              The Complete Web Developer Course 2.0     True   
1388      19421              Free Beginner Electric Guitar Lessons    False   
3556     473160  Web Design for Web Developers: Build Beautiful...    False   
2233      94430             Learn Javascript & JQuery From Scratch     True   
2886     130064  Practical PHP: Master the Basics and Code Dyna...    False   
2034     364426          JavaScript: Understanding the Weird Parts     True   

     price  num_subscribers  num_reviews  num_lectures           level  \
2230  Free           268923         8629            45      All Levels   
776   Free           161029          279            27  Beginner Level   
3385   200           121584        27445           342      All Levels   
640   Free           120291         5924            30  Beginner Level   
3316   200           114512        22412           304      All Levels   
1388  Free           101154         1042            95  Beginner Level   
3556  Free            98867         6512            20      All Levels   
2233    30            84897         2685            10      All Levels   
2886  Free            83737         4598            45      All Levels   
2034   175            79612        16976            85      All Levels   

     content_duration   published_timestamp              subject  
2230       10.5 hours  2013-02-14T07:03:41Z      Web Development  
776         3.5 hours  2013-06-09T15:51:55Z      Web Development  
3385         43 hours  2015-11-02T21:13:27Z      Web Development  
640           3 hours  2014-04-08T16:21:30Z      Web Development  
3316       30.5 hours  2016-03-08T22:28:36Z      Web Development  
1388        4.5 hours  2012-06-15T17:00:33Z  Musical Instruments  
3556          3 hours  2015-04-13T18:29:47Z      Web Development  
2233          2 hours  2013-10-10T15:19:29Z      Web Development  
2886        6.5 hours  2014-07-19T03:02:05Z      Web Development  
2034       11.5 hours  2015-03-12T23:19:54Z      Web Development

#### These are 10 Top Selling Courses.

## 6. Which are Least Selling Courses?

In [12]:
df.sort_values('num_subscribers').head(10)

course_id                                       course_title  is_paid  \
2290    1223240  Learn Pirates of the Caribbean by Ear on the P...     True   
3261    1275872                    The Complete Adobe Spark Course     True   
491     1251582              Best Practices in Corporate Budgeting     True   
484     1032648                               beginning accounting     True   
91      1265814  ABRSM Grade III Piano Class - Handel Sonatina ...     True   
480      993892  Beginners Guitar Course - 'Try it and love it ...     True   
476     1252410  Beginners Guide to Kindle Ebook Covers Design ...     True   
3328     715476        The Cuckoo Strategy to get European Funding     True   
714     1168172       Case studies on credit appraisal for bankers     True   
98      1219820  Accounting - The 6 most important rules you mu...     True   

     price  num_subscribers  num_reviews  num_lectures               level  \
2290    20                0            0             6          All Levels   
3261   150                0            0            29          All Levels   
491     50                0            0             9  Intermediate Level   
484     20                0            0            12  Intermediate Level   
91      20                0            0             5      Beginner Level   
480     25                0            0            16      Beginner Level   
476     20                0            0            16          All Levels   
3328   160                0            0             6          All Levels   
714     20                0            0            15      Beginner Level   
98      20                0            0             9          All Levels   

     content_duration   published_timestamp              subject  
2290          32 mins  2017-05-22T17:14:43Z  Musical Instruments  
3261          2 hours  2017-07-03T17:39:57Z       Graphic Design  
491           2 hours  2017-06-29T22:01:56Z     Business Finance  
484           38 mins  2016-12-26T16:52:47Z     Business Finance  
91            31 mins  2017-07-06T16:12:34Z  Musical Instruments  
480           2 hours  2016-10-31T18:10:06Z  Musical Instruments  
476         1.5 hours  2017-06-19T19:26:46Z       Graphic Design  
3328           1 hour  2016-01-06T02:38:23Z     Business Finance  
714         3.5 hours  2017-05-01T20:16:26Z     Business Finance  
98             1 hour  2017-05-25T21:55:50Z     Business Finance

## 7. Show all courses of Graphic Design where the price is below 100 ?

Here we need to work with price column,but here dtype of price is Object we need to convert it into int, but it has some values as free

In [13]:
df.price.unique()

array(['35', '75', '65', '200', '25', '100', '20', '40', '30', '45',
       '185', '120', '105', '50', 'Free', '145', '55', '85', '110', '150',
       '95', '180', '125', '60', '90', '195', '80', '70', '140', '190',
       '115', '130', '170', '175', '165', '135', '155', '160'],
      dtype=object)

In [14]:
# so we need to replace the Free with 0 

df['price'] = df['price'].replace({'Free':0})

In [15]:
# converting the Object type to int

df.price = df.price.astype(int)

In [16]:
df.subject.unique()

array(['Musical Instruments', 'Business Finance', 'Graphic Design',
       'Web Development'], dtype=object)

In [17]:
graphic_designdf = df.groupby('subject').get_group(('Graphic Design'))
graphic_designdf[:3]

course_id                                       course_title  is_paid  \
8      206592  :::Beginners Guide To Creating Amazing Images ...     True   
20    1165268  15  Motion Graphic Elements in After Effect (S...     True   
35    1164876  3DS MAX - Learn 3Ds MAX tutorial for Beginners...     True   

    price  num_subscribers  num_reviews  num_lectures               level  \
8      25               72            5            17      Beginner Level   
20     30                8            0            20  Intermediate Level   
35     20              785            2            29          All Levels   

   content_duration   published_timestamp         subject  
8            1 hour  2014-04-29T19:50:31Z  Graphic Design  
20          3 hours  2017-04-11T04:27:34Z  Graphic Design  
35        4.5 hours  2017-04-18T16:45:59Z  Graphic Design

In [18]:
graphic_designdf[graphic_designdf.price < 100]

course_id                                       course_title  is_paid  \
8        206592  :::Beginners Guide To Creating Amazing Images ...     True   
20      1165268  15  Motion Graphic Elements in After Effect (S...     True   
35      1164876  3DS MAX - Learn 3Ds MAX tutorial for Beginners...     True   
39       599652                     48 Logo Designs and Techniques     True   
52       794036       8 Stunden professionelles Cinema 4D Training     True   
...         ...                                                ...      ...   
3568     133968                 Web Elements Design With Photoshop     True   
3588    1131780  Whiteboard Animation, Beginner To Expert - In ...     True   
3639    1122306         Wordpress: Comment utiliser le thème FABLE     True   
3671     144222  YOU can Draw, Shade and Paint a Mixed Media Gi...     True   
3681     964478                Zombie Apocalypse Photoshop Actions     True   

      price  num_subscribers  num_reviews  num_lectures               level  \
8        25               72            5            17      Beginner Level   
20       30                8            0            20  Intermediate Level   
35       20              785            2            29          All Levels   
39       20              113            9            48          All Levels   
52       50               18            1            38          All Levels   
...     ...              ...          ...           ...                 ...   
3568     45             5072           38            44          All Levels   
3588     20             1343           25            16          All Levels   
3639     20                1            0            10      Beginner Level   
3671     30             1249           44            19      Beginner Level   
3681     50               12            1            15          All Levels   

     content_duration   published_timestamp         subject  
8              1 hour  2014-04-29T19:50:31Z  Graphic Design  
20            3 hours  2017-04-11T04:27:34Z  Graphic Design  
35          4.5 hours  2017-04-18T16:45:59Z  Graphic Design  
39            9 hours  2015-09-15T00:23:14Z  Graphic Design  
52          8.5 hours  2016-03-22T20:51:28Z  Graphic Design  
...               ...                   ...             ...  
3568          5 hours  2014-10-21T19:18:12Z  Graphic Design  
3588           1 hour  2017-03-14T20:52:30Z  Graphic Design  
3639          39 mins  2017-03-07T16:53:27Z  Graphic Design  
3671          3 hours  2014-01-28T02:30:26Z  Graphic Design  
3681        1.5 hours  2016-09-26T22:19:48Z  Graphic Design  

[484 rows x 11 columns]

# Or we can do it in below way

In [19]:

df[(df.subject == 'Graphic Design') & (df.price<100)]

course_id                                       course_title  is_paid  \
8        206592  :::Beginners Guide To Creating Amazing Images ...     True   
20      1165268  15  Motion Graphic Elements in After Effect (S...     True   
35      1164876  3DS MAX - Learn 3Ds MAX tutorial for Beginners...     True   
39       599652                     48 Logo Designs and Techniques     True   
52       794036       8 Stunden professionelles Cinema 4D Training     True   
...         ...                                                ...      ...   
3568     133968                 Web Elements Design With Photoshop     True   
3588    1131780  Whiteboard Animation, Beginner To Expert - In ...     True   
3639    1122306         Wordpress: Comment utiliser le thème FABLE     True   
3671     144222  YOU can Draw, Shade and Paint a Mixed Media Gi...     True   
3681     964478                Zombie Apocalypse Photoshop Actions     True   

      price  num_subscribers  num_reviews  num_lectures               level  \
8        25               72            5            17      Beginner Level   
20       30                8            0            20  Intermediate Level   
35       20              785            2            29          All Levels   
39       20              113            9            48          All Levels   
52       50               18            1            38          All Levels   
...     ...              ...          ...           ...                 ...   
3568     45             5072           38            44          All Levels   
3588     20             1343           25            16          All Levels   
3639     20                1            0            10      Beginner Level   
3671     30             1249           44            19      Beginner Level   
3681     50               12            1            15          All Levels   

     content_duration   published_timestamp         subject  
8              1 hour  2014-04-29T19:50:31Z  Graphic Design  
20            3 hours  2017-04-11T04:27:34Z  Graphic Design  
35          4.5 hours  2017-04-18T16:45:59Z  Graphic Design  
39            9 hours  2015-09-15T00:23:14Z  Graphic Design  
52          8.5 hours  2016-03-22T20:51:28Z  Graphic Design  
...               ...                   ...             ...  
3568          5 hours  2014-10-21T19:18:12Z  Graphic Design  
3588           1 hour  2017-03-14T20:52:30Z  Graphic Design  
3639          39 mins  2017-03-07T16:53:27Z  Graphic Design  
3671          3 hours  2014-01-28T02:30:26Z  Graphic Design  
3681        1.5 hours  2016-09-26T22:19:48Z  Graphic Design  

[484 rows x 11 columns]

## 8. List out all the courses that are related with 'Python'.

In [20]:
for title in df.course_title:
    if 'Python' in title:
        print(title)

Advanced Scalable Python Web Development Using Flask
Coding for Entrepreneurs: Learn Python, Django, and More.
Complete Python Web Course: Build 8 Python Web Apps
Core: A Web App Reference Guide for Django, Python, and More
Fun and creative web engineering with Python and Web2py
Introduction to QGIS Python Programming
Investment Portfolio Analysis with Python
Learn Python and Django: Payment Processing
Learn Python Django - A Hands-On Course
Learn to code in Python and learn Adobe Photoshop today!
Professional Python Web Development Using Flask
Professional RESTful API Design using Python Flask
Projects in Django and Python
Python Algo Stock Trading: Automate Your Trading!
Python Algo Trading: FX Trading with Oanda
Python Algo Trading: Sentiment Trading with News
Python and Django Full Stack Web Developer Bootcamp
Python for Beginners: Python Programming Language | Tutorial
Python for Finance: Investment Fundamentals & Data Analytics
Python for Trading & Investing
Python Programming: C

In [21]:
df[df.course_title.str.contains('Python')]['course_title']

216     Advanced Scalable Python Web Development Using...
777     Coding for Entrepreneurs: Learn Python, Django...
814     Complete Python Web Course: Build 8 Python Web...
842     Core: A Web App Reference Guide for Django, Py...
1400    Fun and creative web engineering with Python a...
1932              Introduction to QGIS Python Programming
1969            Investment Portfolio Analysis with Python
2294          Learn Python and Django: Payment Processing
2295              Learn Python Django - A Hands-On Course
2341    Learn to code in Python and learn Adobe Photos...
2906      Professional Python Web Development Using Flask
2908    Professional RESTful API Design using Python F...
2929                        Projects in Django and Python
2952    Python Algo Stock Trading: Automate Your Trading!
2953           Python Algo Trading: FX Trading with Oanda
2954     Python Algo Trading: Sentiment Trading with News
2955    Python and Django Full Stack Web Developer Boo...
2956    Python

## 9. What are courses that published in year 2015 ?

In [22]:
df.dtypes

course_id               int64
course_title           object
is_paid                  bool
price                   int32
num_subscribers         int64
num_reviews             int64
num_lectures            int64
level                  object
content_duration       object
published_timestamp    object
subject                object
dtype: object

In [23]:
df.published_timestamp = pd.to_datetime(df.published_timestamp)

In [24]:
# we want only year part
df['year'] = df.published_timestamp.dt.year

In [25]:
df[df['year'] == 2015]

course_id                                       course_title  is_paid  \
10       591880  1 - Concepts of Statistics For Beginners Step ...     True   
16       302450        10 Numbers Every Business Owner Should Know     True   
19       384928  101 Blues riffs - learn how the harmonica supe...     True   
23       550842  16 Guitar Chords to Jam With - (Beginner - Int...     True   
26       486240      2 Easy Steps To Investment And Avoiding Traps     True   
...         ...                                                ...      ...   
3675     551170                         Your First Course on Piano     True   
3676     498488  Your First Successful Forex Trades - With Case...     True   
3677     328960  Your Own Site in 45 Min: The Complete Wordpres...     True   
3678     552700    Your Second Course on Piano: Two Handed Playing     True   
3679     631754  Zend Framework 2: Learn the PHP framework ZF2 ...     True   

      price  num_subscribers  num_reviews  num_lectures               level  \
10      200              273            4            15      Beginner Level   
16       20               13            1             9          All Levels   
19      200             1350           65            55  Intermediate Level   
23       20             1224           19            20      Beginner Level   
26       20              828            1            20          All Levels   
...     ...              ...          ...           ...                 ...   
3675     70             2002           34            33      Beginner Level   
3676    200             1079           34            16          All Levels   
3677    120             1566           29            36          All Levels   
3678     70             1018           12            22      Beginner Level   
3679     40              723          130            37          All Levels   

     content_duration       published_timestamp              subject  year  
10            31 mins 2015-08-30 22:48:34+00:00     Business Finance  2015  
16             1 hour 2015-03-08 19:11:24+00:00     Business Finance  2015  
19          6.5 hours 2015-01-04 21:14:31+00:00  Musical Instruments  2015  
23             1 hour 2015-07-10 19:53:56+00:00  Musical Instruments  2015  
26             1 hour 2015-04-27 23:18:59+00:00     Business Finance  2015  
...               ...                       ...                  ...   ...  
3675          4 hours 2015-07-16 17:07:49+00:00  Musical Instruments  2015  
3676        2.5 hours 2015-05-26 20:48:48+00:00     Business Finance  2015  
3677          4 hours 2015-04-20 22:15:17+00:00      Web Development  2015  
3678          5 hours 2015-10-26 20:04:21+00:00  Musical Instruments  2015  
3679        6.5 hours 2015-11-11 18:55:45+00:00      Web Development  2015  

[1014 rows x 12 columns]

## 10. What are the Max. Number of Subscribers for Each Level of courses ?

In [26]:
df.groupby('level')['num_subscribers'].max()

level
All Levels            268923
Beginner Level        161029
Expert Level            5172
Intermediate Level     29167
Name: num_subscribers, dtype: int64

In [27]:
df.groupby('level').max()

course_id  \
level                           
All Levels            1277924   
Beginner Level        1282064   
Expert Level          1275790   
Intermediate Level    1276182   

                                                         course_title  \
level                                                                   
All Levels                              ６時間でインターバンク市場を攻略！最短距離でトレード基礎力   
Beginner Level             ７日でマスター  ビギナー向け Adobe Illustrator　どきどきセミナー   
Expert Level        [Value Investing] Where White People Keep Thei...   
Intermediate Level                           株式投資で本当のテクニカル分析ができるようになる   

                    is_paid  price  num_subscribers  num_reviews  \
level                                                              
All Levels             True    200           268923        27445   
Beginner Level         True    200           161029         5924   
Expert Level           True    200             5172          249   
Intermediate Level     True    200            29167         3326   

                    num_lectures content_duration       published_timestamp  \
level                                                                         
All Levels                   544        9.5 hours 2017-07-06 21:16:13+00:00   
Beginner Level               779        9.5 hours 2017-07-06 21:46:30+00:00   
Expert Level                 157          9 hours 2017-07-04 18:08:01+00:00   
Intermediate Level           320        9.5 hours 2017-07-05 04:41:54+00:00   

                            subject  year  
level                                      
All Levels          Web Development  2017  
Beginner Level      Web Development  2017  
Expert Level        Web Development  2017  
Intermediate Level  Web Development  2017